src/wasm/extract.cpp

```diff
@@ -917,28 +930,61 @@ std::string extractor::extract(const std::string& list_json) {
       }
 
       uint64_t offset = 0;
+      uint64_t offset_start = 0;
       for (const auto& location : chunk.second) { // 1 chunk => n files
-        const stream::file& file = location.first;
         const std::vector<output_location>& output_locations = files_for_location[location.second];
-
-        offset = seek_input_stream(chunk_source.get(), file, offset);
-
         crypto::checksum checksum;
-        auto input = stream::file_reader::get(*chunk_source, file, &checksum);
         auto outputs = open_outputs(output_locations, output_dir);
 
-        uint64_t output_size = copy_data(input, outputs);
+        wasm_debug("n_outputs = " << outputs.size());
+        if (outputs.size() > 1) {
+          for (auto output : outputs) {
+            std::cout << "multi output: " << output->path() << std::endl;
+          }
+        }
+
+        stream::file file;
+        stream::file_reader::pointer input;
+
+        uint64_t output_size;
+        for (auto output : outputs) {
+          file = location.first;
 
-        if (aborted) {
-          // copy_data is the most likely function to be in while execution is being
-          // aborted
+          wasm_debug("output: " << output->path());
 
-          log_info << "Extraction aborted";
-          abort_zip();
+          if (outputs.size() > 1) {
+            if (output == outputs[0]) {
+              // Save original file offset in chunk
+              offset_start = offset;
+            } else {
+              // Reload chunk containing the duplicated file and its offset
+              chunk_source = stream::chunk_reader::get(*slice_reader, chunk.first, "");
+              offset = offset_start;
+            }
+          }
 
-          json ret;
-          ret["status"] = "Aborted by user";
-          return ret.dump();
+          wasm_debug("pre-copy chunk=" << chunk.first.first_slice << " " << chunk.first.last_slice
```

Comment:
I would rather consider using different format of logging function like:
`wasm_log << "pre-copy chunk=" ....`
or
`conditional_log(extractor::get().get_settings().debug_messages_enabled) << "pre-copy chunk=" ....`

Would be more readable in my opinion.

Reply:
This function (or more specifically macro) was meant to be used only for debug and it's only used in wasm-related code, so I don't really see the point of making this more general or rewriting to be used like a stream.
It might be a good idea to make it more readable, but I assumed it's good enough for now and hoped that it could be addressed later, as important issues should resolved first and without much delay.